### PR TITLE
refactor: add parsers to new filtering code

### PR DESCRIPTION
**Pull request by jake-bassett**

### CODE REVIEW

File: projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import { FilterAttribute } from '../../filter-attribute';
+import { FilterAttributeType } from '../../filter-attribute-type';
+import { FilterOperator } from '../../filter-operators';
+import { ParsedFilter } from '../parsed-filter';
+
+export abstract class AbstractFilterParser<TValue> {
+  public static supportedAttributeTypes: FilterAttributeType[];
+  public static supportedOperators: FilterOperator[];
+
+  protected abstract supportedAttributeTypes(): FilterAttributeType[];
```
Comment (aaron-steinfeld) [inline projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts:10]: Seems strange to have this both as an instance method and static property.

File: projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      .find(op => filterString.includes(op));
+
+    if (matchingOperator === undefined) {
+      return undefined;
+    }
+
+    const parts = filterString
+      .split(matchingOperator)
+      .map(str => str.trim())
+      .filter((part: string | undefined) => (part ?? '').length > 0);
```
Comment (aaron-steinfeld) [inline projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts:54]: how could this be undefined? we invoke trim the line above, so we'd have already errored out

File: projects/components/src/filtering/filter/parser/types/comparison-filter-parser.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
+import { assertUnreachable } from '@hypertrace/common';
+import { isBoolean } from 'lodash-es';
+import { FilterAttribute } from '../../filter-attribute';
+import { FilterAttributeType } from '../../filter-attribute-type';
+import { FilterOperator } from '../../filter-operators';
+import { FilterParser } from '../filter-parser.decorator';
+import { AbstractFilterParser } from './abstract-filter-parser';
+
+@FilterParser({
+  supportedAttributeTypes: [FilterAttributeType.Boolean, FilterAttributeType.Number, FilterAttributeType.String],
```
Comment (aaron-steinfeld) [inline projects/components/src/filtering/filter/parser/types/comparison-filter-parser.ts:10]: How do we match parsers? Not all of these attribute types work with all of these operators.

Comment (jake-bassett) [reply projects/components/src/filtering/filter/parser/types/comparison-filter-parser.ts:10]: Actually going through that part of the code now as I build the lookup service. My thought at the moment is to do it by operator.

File: projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts (hunk 1 above)
Comment (jake-bassett) [reply projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts:10]: This is something I'm not thrilled about either. Its kind of confusing to the reader. The statics in the abstract class are never set. They are strictly there for the decorator to populate for the subclasses.

File: projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts (hunk 2 above)
Comment (jake-bassett) [reply projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts:54]: Ah, true. I'll fix locally for now unless you want this addressed before merge.

Comment (aaron-steinfeld) [reply projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts:54]: No need

File: projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts (hunk 1 above)
Comment (aaron-steinfeld) [reply projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts:10]: I see in this change how they're used at an instance level - are they also being used at a static level? What's being done with the decorater-set values?

File: projects/components/src/filtering/filter/parser/types/comparison-filter-parser.ts (hunk 3 above)
Comment (aaron-steinfeld) [reply projects/components/src/filtering/filter/parser/types/comparison-filter-parser.ts:10]: hmm - we can merge this as is, but may have questions once we get to the lookup part. Curious how the static values are used.

File: projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts (hunk 1 above)
Comment (jake-bassett) [reply projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts:10]: You might be on to something here. This whole statics model with the builders (and now parsers) may not be needed anymore with the way I'm doing the lookups this time. I carried over the statics from the old code, but I don't think I'm going to need them.

Can we keep them for now, and once I've done the full end to end and determined I don't need them anymore I'll remove them?

Comment (jake-bassett) [reply projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts:10]: Take that back. They weren't even in the old code. I'm 99% sure I won't need them. I'll remove them next PR.

Comment (jake-bassett) [reply projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts:10]: Now that I'm looking at how I'm doing the lookup service, I won't actually need those. I'll remove them. Is it okay to do so in the next PR?

**Pull request closed without merging**